// @pnpm/os.env.path-extender/dist/path-extender.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderWindowsReport = exports.addDirToEnvPath = void 0;
const os_env_path_extender_posix_1 = require("@pnpm/os.env.path-extender-posix");
const os_env_path_extender_windows_1 = require("@pnpm/os.env.path-extender-windows");
function addDirToEnvPath(dir, opts) {
    return __awaiter(this, void 0, void 0, function* () {
        if (process.platform === 'win32') {
            return renderWindowsReport(yield (0, os_env_path_extender_windows_1.addDirToWindowsEnvPath)(dir, {
                position: opts.position,
                proxyVarName: opts.proxyVarName,
                overwriteProxyVar: opts.overwrite,
            }));
        }
        return (0, os_env_path_extender_posix_1.addDirToPosixEnvPath)(dir, opts);
    });
}
exports.addDirToEnvPath = addDirToEnvPath;
function renderWindowsReport(changedEnvVariables) {
    const oldSettings = [];
    const newSettings = [];
    for (const changedEnvVariable of changedEnvVariables) {
        if (changedEnvVariable.oldValue) {
            oldSettings.push(`${changedEnvVariable.variable}=${changedEnvVariable.oldValue}`);
        }
        if (changedEnvVariable.newValue) {
            newSettings.push(`${changedEnvVariable.variable}=${changedEnvVariable.newValue}`);
        }
    }
    return {
        oldSettings: oldSettings.join('\n'),
        newSettings: newSettings.join('\n'),
    };
}
exports.renderWindowsReport = renderWindowsReport;
