// @teambit/bvm.list/dist/version-list/version-list.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.VersionList = void 0;
const semver_1 = __importDefault(require("semver"));
class VersionList {
    constructor(entries) {
        this.entries = entries;
    }
    ;
    latest() {
        const sorted = this.sortBySemver('desc');
        return sorted.entries[0];
    }
    stable() {
        const sorted = this.sortBySemver('desc');
        return sorted.entries.find(version => version.releaseType === 'stable');
    }
    find(version) {
        return this.entries.find(entry => entry.version === version);
    }
    toVersionsStringArray() {
        return this.entries.map(entry => entry.version);
    }
    toMap() {
        return new Map(this.entries.map(e => [e.version, e.releaseType]));
    }
    sortBySemver(order = 'desc') {
        let sorted = VersionList.sortList(this.entries, order);
        return new VersionList(sorted);
    }
    static sortList(list, order = 'desc') {
        let sorted;
        if (order === 'asc') {
            sorted = list.sort((entry1, entry2) => {
                return semver_1.default.compare(entry1.version, entry2.version);
            });
        }
        else {
            sorted = list.sort((entry1, entry2) => {
                return semver_1.default.rcompare(entry1.version, entry2.version);
            });
        }
        return sorted;
    }
}
exports.VersionList = VersionList;
